server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/model/Filter.java

```diff
@@ -19,11 +19,44 @@
 
 package org.apache.james.jmap.draft.model;
 
+import java.util.List;
+import java.util.stream.Stream;
+
 import org.apache.james.jmap.draft.json.FilterDeserializer;
 
 import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.github.steveash.guavate.Guavate;
 
 @JsonDeserialize(using = FilterDeserializer.class)
 public interface Filter {
+    class TooDeepFilterHierarchyException extends IllegalArgumentException {
+        TooDeepFilterHierarchyException() {
+            super("Filter depth is higher than maximum allowed value " + MAX_FILTER_DEPTH);
+        }
+    }
+
+    int MAX_FILTER_DEPTH = 10;
+
     String prettyPrint(String indentation);
+
+    default List<FilterCondition> flatten() {
```

Comment:
`flatten` is usually not recursive. You probably mean `visit` because it's a tree. Also you have to define if it's depth-first or breadth-first traversal

Reply:
https://github.com/linagora/james-project/pull/3427 addresses this.